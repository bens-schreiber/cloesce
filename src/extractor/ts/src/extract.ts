import path from "node:path";
import {
  Project,
  Type,
  SourceFile,
  PropertyDeclaration,
  MethodDeclaration,
  SyntaxKind,
  ClassDeclaration,
  Decorator,
} from "ts-morph";

const HTTP_VERBS = ["GET", "POST", "PUT", "PATCH", "DELETE"];

// Mirrors the rust bindings
type CidlType =
  | "Integer"
  | "Real"
  | "Text"
  | "Blob"
  | "D1Database"
  | { Model: string }
  | { Array: CidlType }
  | { HttpResult: CidlType | null };

enum AttributeDecoratorKind {
  PrimaryKey = "PrimaryKey",
  ForeignKey = "ForeignKey",
  OneToOne = "OneToOne",
  OneToMany = "OneToMany",
  ManyToMany = "ManyToMany",
  DataSource = "DataSource",
}

export class CidlExtractor {
  constructor(
    public projectName: string,
    public version: string
  ) {}

  extract(project: Project) {
    const models = project.getSourceFiles().flatMap((sourceFile) => {
      return sourceFile
        .getClasses()
        .filter((classDecl) => hasDecorator(classDecl, "D1"))
        .map((classDecl) => CidlExtractor.model(classDecl, sourceFile));
    });

    return {
      version: this.version,
      project_name: this.projectName,
      language: "TypeScript",
      models,
    };
  }

  private static model(classDecl: ClassDeclaration, sourceFile: SourceFile) {
    const className = classDecl.getName() ?? "<anonymous>";
    const attributes: any[] = [];
    const navigationProperties: any[] = [];
    const dataSources: any[] = [];

    for (const prop of classDecl.getProperties()) {
      const decorators = prop.getDecorators();

      // No decorators means this is a standard attribute
      if (decorators.length === 0) {
        let [cidl_type, nullable] = CidlExtractor.cidlType(prop.getType());
        attributes.push({
          is_primary_key: false,
          foreign_key_reference: null,
          value: {
            name: prop.getName(),
            cidl_type,
            nullable,
          },
        });
        continue;
      }

      // TODO: Limiting to one decorator. Can't get too fancy on us.
      const decorator = decorators[0];
      const name = getDecoratorName(decorator);
      switch (name) {
        case AttributeDecoratorKind.PrimaryKey: {
          let [cidl_type, nullable] = CidlExtractor.cidlType(prop.getType());
          attributes.push({
            is_primary_key: true,
            foreign_key_reference: null,
            value: {
              name: prop.getName(),
              cidl_type,
              nullable,
            },
          });
          break;
        }
        case AttributeDecoratorKind.ForeignKey: {
          let [cidl_type, nullable] = CidlExtractor.cidlType(prop.getType());
          attributes.push({
            is_primary_key: false,
            foreign_key_reference: getDecoratorArgument(decorator, 0),
            value: {
              name: prop.getName(),
              cidl_type,
              nullable,
            },
          });
          break;
        }
        case AttributeDecoratorKind.OneToOne: {
          const reference = getDecoratorArgument(decorator, 0);
          if (!reference) return;

          let [cidl_type, nullable] = CidlExtractor.cidlType(prop.getType());
          navigationProperties.push({
            value: {
              name: prop.getName(),
              cidl_type,
              nullable,
            },
            kind: { [name]: { reference } },
          });
          break;
        }
        case AttributeDecoratorKind.OneToMany:
        case AttributeDecoratorKind.ManyToMany: {
          const reference = getDecoratorArgument(decorator, 0);
          if (!reference) return;

          let [cidl_type, nullable] = CidlExtractor.cidlType(prop.getType());
          navigationProperties.push({
            value: {
              name: prop.getName(),
              cidl_type,
              nullable,
            },
            kind: { [name]: { reference } },
          });
          break;
        }
        case AttributeDecoratorKind.DataSource: {
          const initializer = (prop as any).getInitializer?.();
          const tree = initializer
            ? CidlExtractor.includeTree(initializer, classDecl, sourceFile)
            : [];
          dataSources.push({ name: prop.getName(), tree });
          break;
        }
      }
    }

    const methods = classDecl
      .getMethods()
      .map((m) => CidlExtractor.method(m, sourceFile));

    return {
      name: className,
      attributes,
      navigation_properties: navigationProperties,
      methods,
      data_sources: dataSources,
      source_path: sourceFile.getFilePath().toString(),
    };
  }

  /// Returns a `CidlType` from a TypeScript type, along with if the base value is nullable.
  /// Throws an error if no type can be extracted.
  private static cidlType(type: Type): [CidlType, boolean] {
    let map: Record<string, CidlType> = {
      number: "Integer", // TODO: It's wrong to assume number is always an int.
      string: "Text",
      boolean: "Integer",
      Date: "Text",
      D1Database: "D1Database",
    };

    // TODO: We don't support type unions like Foo | Bar, should we?
    let nullable = type.getUnionTypes().find((t) => t.isNull()) !== undefined;

    // Split by generics and imports
    let split = type.getText().split(/<|>|import\([^)]+\)\.?/);

    let cidlType = split.reduceRight<CidlType | undefined>((acc, x) => {
      // Strip unions
      let base = x
        .split("|")
        .map((s) => s.trim())
        .find((s) => s !== "null" && s !== "undefined")!;

      // Disregard any promises, they have no meaning as of now
      if (!base || base === "Promise") return acc!;

      // Primitive or nullable primitive
      if (map[base] !== undefined) return map[base];

      // Array of primitive
      if (base.endsWith("[]")) {
        const item = base.slice(0, -2);
        return map[item] !== undefined
          ? { Array: map[item] }
          : { Array: { Model: item } };
      }

      // Skip void
      if (base == "void") return acc;

      // Result wrapper
      if (base === "Result") {
        return { HttpResult: acc == undefined ? null : acc };
      }

      return { Model: base };
    }, undefined)!;

    return [cidlType, nullable];
  }

  private static includeTree(
    obj: any,
    currentClass: ClassDeclaration,
    sf: SourceFile
  ): any[] {
    if (!obj.isKind || !obj.isKind(SyntaxKind.ObjectLiteralExpression)) {
      return [];
    }

    const result: any[] = [];
    for (const prop of obj.getProperties()) {
      if (!prop.isKind(SyntaxKind.PropertyAssignment)) continue;

      let navProp = findPropertyByName(currentClass, prop.getName());
      if (!navProp) {
        console.log(
          `  Warning: Could not find property "${prop.getName()}" in class ${currentClass.getName()}`
        );
        continue;
      }

      let [cidl_type, _] = CidlExtractor.cidlType(navProp.getType());
      if (typeof cidl_type === "string") continue;

      const typedValue = {
        name: navProp.getName(),
        cidl_type,
        nullable: false, // TODO: hardcoding this for now, it doesn't mean anything for the IncludeTree
      };

      // Recurse for nested includes
      const initializer = (prop as any).getInitializer?.();
      let nestedTree: any[] = [];

      if (initializer?.isKind?.(SyntaxKind.ObjectLiteralExpression)) {
        let targetModel = getModelName(cidl_type);
        const targetClass = currentClass
          .getSourceFile()
          .getProject()
          .getSourceFiles()
          .flatMap((f) => f.getClasses())
          .find((c) => c.getName() === targetModel);

        if (targetClass) {
          nestedTree = CidlExtractor.includeTree(initializer, targetClass, sf);
        }
      }

      result.push([typedValue, nestedTree]);
    }

    return result;
  }

  private static method(method: MethodDeclaration, sf: SourceFile): any {
    const decorators = method.getDecorators();
    const decoratorNames = decorators.map((d) => getDecoratorName(d));

    const httpVerb =
      HTTP_VERBS.find((verb) => decoratorNames.includes(verb)) || null;

    const parameters: any[] = [];

    for (const param of method.getParameters()) {
      let [cidl_type, nullable] = CidlExtractor.cidlType(param.getType());
      parameters.push({
        name: param.getName(),
        cidl_type,
        nullable,
      });
    }

    // TODO: return types cant be nullable??
    let [return_type, _] = CidlExtractor.cidlType(method.getReturnType());

    return {
      name: method.getName(),
      is_static: method.isStatic(),
      http_verb: httpVerb,
      return_type,
      parameters,
    };
  }
}

function getDecoratorName(decorator: Decorator): string {
  const name = decorator.getName() ?? decorator.getExpression().getText();
  return String(name).replace(/\(.*\)$/, "");
}

function getDecoratorArgument(
  decorator: Decorator,
  index: number
): string | undefined {
  const args = decorator.getArguments();
  if (!args[index]) return undefined;

  const arg = args[index] as any;

  // Identifier
  if (arg.getKind?.() === SyntaxKind.Identifier) {
    return arg.getText();
  }

  // String literal
  const text = arg.getText?.();
  if (!text) return undefined;

  const match = text.match(/^['"](.*)['"]$/);
  return match ? match[1] : text;
}

function getModelName(t: CidlType): string | undefined {
  if (typeof t === "string") return undefined;

  if ("Model" in t) {
    return t.Model;
  } else if ("Array" in t) {
    return getModelName(t.Array);
  } else if ("HttpResult" in t) {
    if (t == null) return undefined;
    return getModelName(t.HttpResult!);
  }

  return undefined;
}

function findPropertyByName(
  cls: ClassDeclaration,
  name: string
): PropertyDeclaration | undefined {
  // Try exact match first
  const exactMatch = cls.getProperties().find((p) => p.getName() === name);
  return exactMatch;
}

function hasDecorator(
  node: { getDecorators(): Decorator[] },
  name: string
): boolean {
  return node.getDecorators().some((d) => {
    const decoratorName = getDecoratorName(d);
    return decoratorName === name || decoratorName.endsWith("." + name);
  });
}
