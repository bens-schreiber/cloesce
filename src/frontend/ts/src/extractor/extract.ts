import {
  Node as MorphNode,
  Project,
  Type,
  SourceFile,
  PropertyDeclaration,
  MethodDeclaration,
  SyntaxKind,
  ClassDeclaration,
  Decorator,
  Expression,
} from "ts-morph";

import {
  CidlIncludeTree,
  CloesceAst,
  CidlType,
  DataSource,
  Either,
  HttpVerb,
  Model,
  ModelAttribute,
  ModelMethod,
  NamedTypedValue,
  NavigationProperty,
  WranglerEnv,
  left,
  right,
  ExtractorError,
  ExtractorErrorCode,
  PlainOldObject,
  CrudKind,
} from "../common.js";
import { TypeFormatFlags } from "typescript";

enum AttributeDecoratorKind {
  PrimaryKey = "PrimaryKey",
  ForeignKey = "ForeignKey",
  OneToOne = "OneToOne",
  OneToMany = "OneToMany",
  ManyToMany = "ManyToMany",
  DataSource = "DataSource",
}

enum ClassDecoratorKind {
  D1 = "D1",
  WranglerEnv = "WranglerEnv",
  PlainOldObject = "PlainOldObject",
}

enum ParameterDecoratorKind {
  Inject = "Inject",
}

export class CidlExtractor {
  constructor(
    public projectName: string,
    public version: string
  ) {}

  extract(project: Project): Either<ExtractorError, CloesceAst> {
    const models: Record<string, Model> = {};
    const poos: Record<string, PlainOldObject> = {};
    const wranglerEnvs: WranglerEnv[] = [];
    let app_source: string | null = null;

    for (const sourceFile of project.getSourceFiles()) {
      if (
        sourceFile.getBaseName() === "app.cloesce.ts" ||
        sourceFile.getBaseName() === "seed__app.cloesce.ts" // hardcoding for tests
      ) {
        const app = CidlExtractor.app(sourceFile);
        if (!app.ok) {
          return app;
        }

        app_source = app.value;
      }

      for (const classDecl of sourceFile.getClasses()) {
        if (hasDecorator(classDecl, ClassDecoratorKind.D1)) {
          const result = CidlExtractor.model(classDecl, sourceFile);

          // Error: propogate from models
          if (!result.ok) {
            result.value.addContext((old) => `${classDecl.getName()}.${old}`);
            return result;
          }
          models[result.value.name] = result.value;
          continue;
        }

        if (hasDecorator(classDecl, ClassDecoratorKind.PlainOldObject)) {
          const result = CidlExtractor.poo(classDecl, sourceFile);

          // Error: propogate from models
          if (!result.ok) {
            result.value.addContext((old) => `${classDecl.getName()}.${old}`);
            return result;
          }
          poos[result.value.name] = result.value;
          continue;
        }

        if (hasDecorator(classDecl, ClassDecoratorKind.WranglerEnv)) {
          wranglerEnvs.push({
            name: classDecl.getName(),
            source_path: sourceFile.getFilePath().toString(),
          } as WranglerEnv);
        }
      }
    }

    // Error: A wrangler environment is required
    if (wranglerEnvs.length < 1) {
      return err(ExtractorErrorCode.MissingWranglerEnv);
    }

    // Error: Only one wrangler environment can exist
    if (wranglerEnvs.length > 1) {
      return err(
        ExtractorErrorCode.TooManyWranglerEnvs,
        (e) => (e.context = wranglerEnvs.map((w) => w.name).toString())
      );
    }

    return right({
      version: this.version,
      project_name: this.projectName,
      language: "TypeScript",
      wrangler_env: wranglerEnvs[0],
      models,
      poos,
      app_source,
    });
  }

  private static app(sourceFile: SourceFile): Either<ExtractorError, string> {
    const symbol = sourceFile.getDefaultExportSymbol();
    const decl = symbol?.getDeclarations()[0];

    if (!decl) {
      return err(ExtractorErrorCode.AppMissingDefaultExport);
    }

    const getTypeText = (): string | undefined => {
      if (MorphNode.isExportAssignment(decl)) {
        return decl.getExpression()?.getType().getText();
      }
      if (MorphNode.isVariableDeclaration(decl)) {
        return decl.getInitializer()?.getType().getText();
      }
      return undefined;
    };

    const typeText = getTypeText();
    if (typeText === "CloesceApp") {
      return right(sourceFile.getFilePath().toString());
    }

    return err(ExtractorErrorCode.AppMissingDefaultExport);
  }

  private static model(
    classDecl: ClassDeclaration,
    sourceFile: SourceFile
  ): Either<ExtractorError, Model> {
    const name = classDecl.getName()!;
    const attributes: ModelAttribute[] = [];
    const navigationProperties: NavigationProperty[] = [];
    const dataSources: Record<string, DataSource> = {};
    const methods: Record<string, ModelMethod> = {};
    let cruds: CrudKind[] = [];
    let primary_key: NamedTypedValue | undefined = undefined;

    // Extract crud methods
    const crudDecorator = classDecl
      .getDecorators()
      .find((d) => getDecoratorName(d) === "CRUD");
    if (crudDecorator) {
      cruds = getCrudKinds(crudDecorator);
    }

    // Iterate attribtutes
    for (const prop of classDecl.getProperties()) {
      const decorators = prop.getDecorators();
      const typeRes = CidlExtractor.cidlType(prop.getType());

      // Error: invalid property type
      if (!typeRes.ok) {
        typeRes.value.context = prop.getName();
        typeRes.value.snippet = prop.getText();
        return typeRes;
      }

      // No decorators means this is a standard attribute
      if (decorators.length === 0) {
        const cidl_type = typeRes.value;
        attributes.push({
          foreign_key_reference: null,
          value: {
            name: prop.getName(),
            cidl_type,
          },
        });
        continue;
      }

      // TODO: Limiting to one decorator. Can't get too fancy on us.
      const decorator = decorators[0];
      const name = getDecoratorName(decorator);

      // Process decorators
      const cidl_type = typeRes.value;
      switch (name) {
        case AttributeDecoratorKind.PrimaryKey: {
          primary_key = {
            name: prop.getName(),
            cidl_type,
          };
          break;
        }
        case AttributeDecoratorKind.ForeignKey: {
          attributes.push({
            foreign_key_reference: getDecoratorArgument(decorator, 0) ?? null,
            value: {
              name: prop.getName(),
              cidl_type,
            },
          });
          break;
        }
        case AttributeDecoratorKind.OneToOne: {
          const reference = getDecoratorArgument(decorator, 0);

          // Error: One to one navigation properties requre a reference
          if (!reference) {
            return err(
              ExtractorErrorCode.MissingNavigationPropertyReference,
              (e) => {
                e.snippet = prop.getText();
                e.context = prop.getName();
              }
            );
          }

          let model_name = getObjectName(cidl_type);

          // Error: navigation properties require a model reference
          if (!model_name) {
            return err(
              ExtractorErrorCode.MissingNavigationPropertyReference,
              (e) => {
                e.snippet = prop.getText();
                e.context = prop.getName();
              }
            );
          }

          navigationProperties.push({
            var_name: prop.getName(),
            model_name,
            kind: { OneToOne: { reference } },
          });
          break;
        }
        case AttributeDecoratorKind.OneToMany: {
          const reference = getDecoratorArgument(decorator, 0);
          // Error: One to one navigation properties requre a reference
          if (!reference) {
            return err(
              ExtractorErrorCode.MissingNavigationPropertyReference,
              (e) => {
                e.snippet = prop.getText();
                e.context = prop.getName();
              }
            );
          }

          let model_name = getObjectName(cidl_type);

          // Error: navigation properties require a model reference
          if (!model_name) {
            return err(
              ExtractorErrorCode.MissingNavigationPropertyReference,
              (e) => {
                e.snippet = prop.getText();
                e.context = prop.getName();
              }
            );
          }

          navigationProperties.push({
            var_name: prop.getName(),
            model_name,
            kind: { OneToMany: { reference } },
          });
          break;
        }
        case AttributeDecoratorKind.ManyToMany: {
          const unique_id = getDecoratorArgument(decorator, 0);

          // Error: many to many attribtues require a unique id
          if (!unique_id)
            return err(ExtractorErrorCode.MissingManyToManyUniqueId, (e) => {
              e.snippet = prop.getText();
              e.context = prop.getName();
            });

          // Error: navigation properties require a model reference
          let model_name = getObjectName(cidl_type);
          if (!model_name) {
            return err(
              ExtractorErrorCode.MissingNavigationPropertyReference,
              (e) => {
                e.snippet = prop.getText();
                e.context = prop.getName();
              }
            );
          }

          navigationProperties.push({
            var_name: prop.getName(),
            model_name,
            kind: { ManyToMany: { unique_id } },
          });
          break;
        }
        case AttributeDecoratorKind.DataSource: {
          const initializer = prop.getInitializer();
          const treeRes = CidlExtractor.includeTree(
            initializer,
            classDecl,
            sourceFile
          );

          if (!treeRes.ok) {
            treeRes.value.addContext((old) => `${prop.getName()} ${old}`);
            treeRes.value.snippet = prop.getText();
            return treeRes;
          }

          dataSources[prop.getName()] = {
            name: prop.getName(),
            tree: treeRes.value,
          };
          break;
        }
      }
    }

    if (primary_key == undefined) {
      return err(ExtractorErrorCode.MissingPrimaryKey, (e) => {
        e.snippet = classDecl.getText();
      });
    }

    // Process methods
    for (const m of classDecl.getMethods()) {
      const result = CidlExtractor.method(m);
      if (!result.ok) {
        result.value.addContext((old) => `${m.getName()} ${old}`);
        return left(result.value);
      }
      methods[result.value.name] = result.value;
    }

    return right({
      name,
      attributes,
      primary_key,
      navigation_properties: navigationProperties,
      methods,
      cruds,
      data_sources: dataSources,
      source_path: sourceFile.getFilePath().toString(),
    });
  }

  private static poo(
    classDecl: ClassDeclaration,
    sourceFile: SourceFile
  ): Either<ExtractorError, PlainOldObject> {
    const name = classDecl.getName()!;
    const attributes: NamedTypedValue[] = [];

    for (const prop of classDecl.getProperties()) {
      const typeRes = CidlExtractor.cidlType(prop.getType());

      // Error: invalid property type
      if (!typeRes.ok) {
        typeRes.value.context = prop.getName();
        typeRes.value.snippet = prop.getText();
        return typeRes;
      }

      const cidl_type = typeRes.value;
      attributes.push({
        name: prop.getName(),
        cidl_type,
      });
      continue;
    }

    return right({
      name,
      attributes,
      source_path: sourceFile.getFilePath().toString(),
    });
  }

  private static readonly primTypeMap: Record<string, CidlType> = {
    number: "Integer",
    Number: "Integer",
    string: "Text",
    String: "Text",
    boolean: "Integer",
    Boolean: "Integer",
    Date: "Text",
  };

  private static cidlType(
    type: Type,
    inject: boolean = false
  ): Either<ExtractorError, CidlType> {
    // Void
    if (type.isVoid()) {
      return right("Void");
    }

    // Null
    if (type.isNull()) {
      return right({ Nullable: "Void" });
    }

    // Nullable via union
    const [unwrappedType, nullable] = unwrapNullable(type);

    const tyText = unwrappedType
      .getText(undefined, TypeFormatFlags.UseAliasDefinedOutsideCurrentScope)
      .split("|")[0]
      .trim();

    // Primitives
    const prim = this.primTypeMap[tyText];
    if (prim) {
      return right(wrapNullable(prim, nullable));
    }

    const generics = [
      ...unwrappedType.getAliasTypeArguments(),
      ...unwrappedType.getTypeArguments(),
    ];

    // Error: can't handle multiple generics
    if (generics.length > 1) {
      return err(ExtractorErrorCode.MultipleGenericType);
    }

    // No generics -> inject or object
    if (generics.length === 0) {
      const base = inject ? { Inject: tyText } : { Object: tyText };
      return right(wrapNullable(base, nullable));
    }

    // Single generic
    const genericTy = generics[0];
    const symbolName = unwrappedType.getSymbol()?.getName();
    const aliasName = unwrappedType.getAliasSymbol()?.getName();

    if (aliasName === "DeepPartial") {
      const [_, genericTyNullable] = unwrapNullable(genericTy);
      const genericTyGenerics = [
        ...genericTy.getAliasTypeArguments(),
        ...genericTy.getTypeArguments(),
      ];

      // Expect partials to be of the exact form DeepPartial<Model>
      if (
        genericTyNullable ||
        genericTy.isUnion() ||
        genericTyGenerics.length > 0
      ) {
        return err(ExtractorErrorCode.InvalidPartialType);
      }

      return right(
        wrapNullable(
          {
            Partial: genericTy
              .getText(
                undefined,
                TypeFormatFlags.UseAliasDefinedOutsideCurrentScope
              )
              .split("|")[0]
              .trim(),
          },
          nullable
        )
      );
    }

    if (symbolName === "Promise" || aliasName === "IncludeTree") {
      return wrapGeneric(genericTy, nullable, (inner) => inner);
    }

    if (unwrappedType.isArray()) {
      return wrapGeneric(genericTy, nullable, (inner) => ({ Array: inner }));
    }

    if (aliasName === "HttpResult") {
      return wrapGeneric(genericTy, nullable, (inner) => ({
        HttpResult: inner,
      }));
    }

    // Error: unknown type
    return err(ExtractorErrorCode.UnknownType);

    function wrapNullable(inner: CidlType, isNullable: boolean): CidlType {
      if (isNullable) {
        return { Nullable: inner };
      } else {
        return inner;
      }
    }

    function wrapGeneric(
      t: Type,
      isNullable: boolean,
      wrapper: (inner: CidlType) => CidlType
    ): Either<ExtractorError, CidlType> {
      const res = CidlExtractor.cidlType(t, inject);

      // Error: propogated from `cidlType`
      if (!res.ok) {
        return res;
      }

      return right(wrapNullable(wrapper(res.value), isNullable));
    }

    function unwrapNullable(ty: Type): [Type, boolean] {
      if (ty.isUnion()) {
        const nonNull = ty.getUnionTypes().filter((t) => !t.isNull());
        if (nonNull.length === 1) {
          return [nonNull[0], true];
        }
      }
      return [ty, false];
    }
  }

  private static includeTree(
    expr: Expression | undefined,
    currentClass: ClassDeclaration,
    sf: SourceFile
  ): Either<ExtractorError, CidlIncludeTree> {
    // Include trees must be of the expected form
    if (
      !expr ||
      !expr.isKind ||
      !expr.isKind(SyntaxKind.ObjectLiteralExpression)
    ) {
      return err(ExtractorErrorCode.InvalidIncludeTree);
    }

    const result: CidlIncludeTree = {};
    for (const prop of expr.getProperties()) {
      if (!prop.isKind(SyntaxKind.PropertyAssignment)) continue;

      // Error: navigation property not found
      const navProp = findPropertyByName(currentClass, prop.getName());
      if (!navProp) {
        return err(
          ExtractorErrorCode.UnknownNavigationPropertyReference,
          (e) => {
            e.snippet = expr.getText();
            e.context = prop.getName();
          }
        );
      }

      const typeRes = CidlExtractor.cidlType(navProp.getType());

      // Error: invalid referenced nav prop type
      if (!typeRes.ok) {
        typeRes.value.snippet = navProp.getText();
        typeRes.value.context = prop.getName();
        return typeRes;
      }

      // Error: invalid referenced nav prop type
      const cidl_type = typeRes.value;
      if (typeof cidl_type === "string") {
        return err(
          ExtractorErrorCode.InvalidNavigationPropertyReference,
          (e) => {
            ((e.snippet = navProp.getText()), (e.context = prop.getName()));
          }
        );
      }

      // Recurse for nested includes
      const initializer = (prop as any).getInitializer?.();
      let nestedTree: CidlIncludeTree = {};
      if (initializer?.isKind?.(SyntaxKind.ObjectLiteralExpression)) {
        const targetModel = getObjectName(cidl_type);
        const targetClass = currentClass
          .getSourceFile()
          .getProject()
          .getSourceFiles()
          .flatMap((f) => f.getClasses())
          .find((c) => c.getName() === targetModel);

        if (targetClass) {
          const treeRes = CidlExtractor.includeTree(
            initializer,
            targetClass,
            sf
          );

          // Error: Propogated from `includeTree`
          if (!treeRes.ok) {
            treeRes.value.snippet = expr.getText();
            return treeRes;
          }

          nestedTree = treeRes.value;
        }
      }

      result[navProp.getName()] = nestedTree;
    }

    return right(result);
  }

  private static method(
    method: MethodDeclaration
  ): Either<ExtractorError, ModelMethod> {
    const decorators = method.getDecorators();
    const decoratorNames = decorators.map((d) => getDecoratorName(d));

    const httpVerb = decoratorNames.find((name) =>
      Object.values(HttpVerb).includes(name as HttpVerb)
    ) as HttpVerb;

    const parameters: NamedTypedValue[] = [];

    for (const param of method.getParameters()) {
      // Handle injected param
      if (param.getDecorator(ParameterDecoratorKind.Inject)) {
        const typeRes = CidlExtractor.cidlType(param.getType(), true);

        // Error: invalid type
        if (!typeRes.ok) {
          typeRes.value.snippet = method.getText();
          typeRes.value.context = param.getName();
          return typeRes;
        }

        parameters.push({
          name: param.getName(),
          cidl_type: typeRes.value,
        });
        continue;
      }

      // Handle all other params
      const typeRes = CidlExtractor.cidlType(param.getType());

      // Error: invalid type
      if (!typeRes.ok) {
        typeRes.value.snippet = method.getText();
        typeRes.value.context = param.getName();
        return typeRes;
      }

      parameters.push({
        name: param.getName(),
        cidl_type: typeRes.value,
      });
    }

    const typeRes = CidlExtractor.cidlType(method.getReturnType());

    // Error: invalid type
    if (!typeRes.ok) {
      typeRes.value.snippet = method.getText();
      return typeRes;
    }

    return right({
      name: method.getName(),
      is_static: method.isStatic(),
      http_verb: httpVerb,
      return_type: typeRes.value,
      parameters,
    });
  }
}

function err(
  code: ExtractorErrorCode,
  fn?: (extractorErr: ExtractorError) => void
): Either<ExtractorError, never> {
  let e = new ExtractorError(code);
  if (fn) {
    fn(e);
  }
  return left(e);
}

function getDecoratorName(decorator: Decorator): string {
  const name = decorator.getName() ?? decorator.getExpression().getText();
  return String(name).replace(/\(.*\)$/, "");
}

function getDecoratorArgument(
  decorator: Decorator,
  index: number
): string | undefined {
  const args = decorator.getArguments();
  if (!args[index]) return undefined;

  const arg = args[index] as any;

  if (arg.getKind?.() === SyntaxKind.Identifier) {
    return arg.getText();
  }

  return arg.getLiteralValue();
}

function getObjectName(t: CidlType): string | undefined {
  if (typeof t === "string") return undefined;

  if ("Object" in t) {
    return t.Object;
  } else if ("Array" in t) {
    return getObjectName(t.Array);
  } else if ("HttpResult" in t) {
    if (t == null) return undefined;
    return getObjectName(t.HttpResult!);
  }

  return undefined;
}

function getCrudKinds(d: Decorator): CrudKind[] {
  const arg = d.getArguments()[0];
  if (!arg) return [];

  if (MorphNode.isArrayLiteralExpression(arg)) {
    return arg
      .getElements()
      .map(
        (e) =>
          (MorphNode.isStringLiteral(e)
            ? e.getLiteralValue()
            : e.getText()) as CrudKind
      );
  }

  return [];
}

function findPropertyByName(
  cls: ClassDeclaration,
  name: string
): PropertyDeclaration | undefined {
  const exactMatch = cls.getProperties().find((p) => p.getName() === name);
  return exactMatch;
}

function hasDecorator(
  node: { getDecorators(): Decorator[] },
  name: string
): boolean {
  return node.getDecorators().some((d) => {
    const decoratorName = getDecoratorName(d);
    return decoratorName === name || decoratorName.endsWith("." + name);
  });
}
