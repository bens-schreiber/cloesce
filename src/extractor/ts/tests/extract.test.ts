import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { extractModels } from "../src/extract.js";
import { readFileSync } from "fs";

import { readdirSync } from "fs";
import { join } from "path";

test("actions snapshot", () => {
  let models = extractModels({
    version: "0.0.2",
    projectName: "actions",
    cwd: "./tests/fixtures",
    tsconfigPath: "./tests/fixtures/tsconfig.json",
  });

  expect(models).toMatchSnapshot();
});
