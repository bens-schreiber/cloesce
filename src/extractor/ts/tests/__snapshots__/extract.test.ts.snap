// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions snapshot 1`] = `
{
  "language": "TypeScript",
  "models": [
    {
      "attributes": [
        {
          "foreign_key": null,
          "primary_key": true,
          "value": {
            "cidl_type": "Integer",
            "name": "id",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Text",
            "name": "name",
            "nullable": false,
          },
        },
      ],
      "data_sources": [],
      "methods": [
        {
          "http_verb": "POST",
          "is_static": true,
          "name": "create",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Text",
              "name": "payload",
              "nullable": false,
            },
          ],
        },
        {
          "http_verb": "PUT",
          "is_static": true,
          "name": "update",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Integer",
              "name": "id",
              "nullable": false,
            },
            {
              "cidl_type": "Text",
              "name": "payload",
              "nullable": true,
            },
          ],
        },
        {
          "http_verb": "PATCH",
          "is_static": true,
          "name": "patch",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Text",
              "name": "phrase",
              "nullable": false,
            },
          ],
        },
        {
          "http_verb": "DELETE",
          "is_static": true,
          "name": "remove",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Integer",
              "name": "id",
              "nullable": false,
            },
          ],
        },
      ],
      "name": "Actions",
      "navigation_properties": [],
      "source_path": "actions.cloesce.ts",
    },
    {
      "attributes": [
        {
          "foreign_key": null,
          "primary_key": true,
          "value": {
            "cidl_type": "Integer",
            "name": "id",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Text",
            "name": "name",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Integer",
            "name": "breed",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Text",
            "name": "preferred_treat",
            "nullable": true,
          },
        },
      ],
      "data_sources": [],
      "methods": [
        {
          "http_verb": "GET",
          "is_static": false,
          "name": "get_name",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
          ],
        },
        {
          "http_verb": "GET",
          "is_static": false,
          "name": "get_breed",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
          ],
        },
        {
          "http_verb": "POST",
          "is_static": true,
          "name": "woof",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Text",
              "name": "phrase",
              "nullable": false,
            },
          ],
        },
      ],
      "name": "Dog",
      "navigation_properties": [],
      "source_path": "dog.cloesce.ts",
    },
    {
      "attributes": [
        {
          "foreign_key": null,
          "primary_key": true,
          "value": {
            "cidl_type": "Integer",
            "name": "id",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Text",
            "name": "name",
            "nullable": false,
          },
        },
        {
          "foreign_key": null,
          "primary_key": false,
          "value": {
            "cidl_type": "Text",
            "name": "middle_name",
            "nullable": true,
          },
        },
      ],
      "data_sources": [],
      "methods": [
        {
          "http_verb": "GET",
          "is_static": false,
          "name": "foo",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
          ],
        },
        {
          "http_verb": "POST",
          "is_static": true,
          "name": "speak",
          "parameters": [
            {
              "cidl_type": "D1Database",
              "name": "db",
              "nullable": false,
            },
            {
              "cidl_type": "Text",
              "name": "phrase",
              "nullable": false,
            },
          ],
        },
      ],
      "name": "Person",
      "navigation_properties": [],
      "source_path": "person.cloesce.ts",
    },
  ],
  "project_name": "actions",
  "version": "0.0.1",
}
`;
